File: src/main/java/cc/duduhuo/git/conflict/InDocumentListener.java
package cc.duduhuo.git.conflict;

import cc.duduhuo.git.conflict.tool.DocumentTools;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.event.DocumentEvent;
import com.intellij.openapi.editor.event.DocumentListener;
import org.jetbrains.annotations.NotNull;

import static cc.duduhuo.git.conflict.Global.sConflictItemMap;
import static cc.duduhuo.git.conflict.Global.sIsHighlightMap;

/**
 * =======================================================
 * Author: liying - liruoer2008@yeah.net
 * Datetime: 2018/4/22 10:37
 * Description:
 * Remarks:
 * =======================================================
 */
public final class InDocumentListener implements DocumentListener {
    private final Editor mEditor;

    public InDocumentListener(Editor editor) {
        this.mEditor = editor;
    }

    @Override
    public void documentChanged(@NotNull DocumentEvent event) {
        boolean hasConflict = DocumentTools.showConflict(mEditor);
        if (!hasConflict) {
            Document document = mEditor.getDocument();
            sIsHighlightMap.put(mEditor, false);
            sConflictItemMap.remove(document);
            InDocumentListener listener = Global.sDocumentListenerMap.get(document);
            if (listener != null) {
                document.removeDocumentListener(listener);
                Global.sDocumentListenerMap.remove(document);
            }
        }
    }
}

File: src/main/java/cc/duduhuo/git/conflict/InDocumentListener.kt
package cc.duduhuo.git.conflict

import cc.duduhuo.git.conflict.tool.DocumentTools
import cc.duduhuo.git.conflict.tool.ext.removeInDocumentListenerIfExist
import com.intellij.openapi.editor.Document
import com.intellij.openapi.editor.event.DocumentEvent
import com.intellij.openapi.editor.event.DocumentListener
import com.intellij.openapi.project.Project

/**
 * =======================================================
 * Author: liying - liruoer2008@yeah.net
 * Datetime: 2018/4/22 10:37
 * Description:
 * Remarks:
 * =======================================================
 */
class InDocumentListener(private val document: Document, private val project: Project) : DocumentListener {

    override fun documentChanged(event: DocumentEvent) {
        // println("documentChanged: ${document}, ${event}")
        // TODO highlight conflicts only when the document has highlighters
        val conflictsCount = DocumentTools.showConflict(document, project)
        if (conflictsCount == 0) {
            // remove document listener
            document.removeInDocumentListenerIfExist()
        }
    }
}
